PyInstaller/depend/bindepend.py

```diff
@@ -457,53 +459,80 @@ def _get_imports_macholib(filename, search_paths):
+                    if rpath.startswith("@") and not rpath.startswith(("@executable_path", "@loader_path")):
+                        logger.warning("Unsupported rpath format %r found in binary %r - ignoring...", rpath, filename)
+                        continue
+                    run_paths.append(rpath)
+        return run_paths
+
+    @functools.lru_cache
+    def get_run_paths_and_referenced_libs(filename):
+        # Walk through Mach-O headers, and collect all referenced libraries and run paths.
+        m = MachO(filename)
+        return _get_referenced_libs(m), _get_run_paths(m)
+
+    @functools.lru_cache
+    def get_run_paths(filename):
+        # Walk through Mach-O headers, and collect only run paths.
+        return _get_run_paths(MachO(filename))
+
+    # Collect referenced libraries and run paths from the input binary.
+    referenced_libs, run_paths = get_run_paths_and_referenced_libs(filename)
+
+    # On macOS, run paths (rpaths) are inherited from the executable that loads the given shared library (or from the
+    # shared library that loads the given shared library). This means that shared libraries and python binary extensions
+    # can reference other shared libraries using @rpath without having set any run paths themselves.
     #
-    # Our analysis does not account for inherited run-paths, and we attempt to work around this limitation by
-    # registering the following fall-back run-path.
-    run_paths.add(os.path.join(compat.base_prefix, 'lib'))
+    # In order to simulate the run path inheritance that happens in unfrozen python programs, we need to augment the
+    # run paths from the given binary with those set by the python interpreter executable (`sys.executable`). Anaconda
+    # python, for example, sets the run path on the python executable to `@loader_path/../lib`, which allows python
+    # extensions to reference shared libraries in the Anaconda environment's `lib` directory via only `@rpath`
+    # (for example, the `_ssl` extension can reference the OpenSSL library as `@rpath/libssl.3.dylib`). In another
+    # example, python executable has its run path set to the top-level directory of its .framework bundle; in this
+    # case the `ssl` extension references the OpenSSL library as `@rpath/Versions/3.10/lib/libssl.1.1.dylib`.
+    run_paths += get_run_paths(python_bin)
+
+    # This fallback should be fully superseded by the above recovery of run paths from python executable; but for now,
+    # keep it around in case of unforeseen corner cases.
+    run_paths.append(os.path.join(compat.base_prefix, 'lib'))
+
+    # De-duplicate run_paths while preserving the order of the list. This is better than using a set for run paths
+    # because the set can do arbriary re-ording and we want to keep the priority in the order we found the paths above.
+    # Using collections.OrderedDict ensures the behavior for older versions of python vs. just using dict directly.
```

Comment:
We support only python >= 3.8, so using `collections.OrderedDict` instead of plain `dict` (which preserves order insertion since 3.7) has no benefit w.r.t. compatibility.